nbbuild/antsrc/org/netbeans/nbbuild/HgId.java

```diff
@@ -53,33 +56,43 @@ public void setProperty(String property) {
         this.property = property;
     }
 
-    public @Override void execute() throws BuildException {
+    @Override
+    public void execute() throws BuildException {
         if (file == null || property == null) {
             throw new BuildException("define file and property");
         }
-        File dirstate = null;
-        for (File root = file; root != null; root = root.getParentFile()) {
-            File ds = new File(new File(root, ".hg"), "dirstate");
-            if (ds.isFile()) {
-                dirstate = ds;
+        Path headroot = null;
+        for (Path root = file.toPath(); root != null; root = root.getParent()) {
+            Path headpath = root.resolve(".git/HEAD");
+            if (Files.isRegularFile(headpath)) {
+                headroot = headpath;
                 break;
             }
         }
         String id = "unknown-revn";
-        if (dirstate != null && dirstate.length() >= 6) {
-            try {
-                try (InputStream is = new FileInputStream(dirstate)) {
-                    byte[] data = new byte[6];
-                    if (is.read(data) < 6) {
-                        throw new IOException("truncated read");
+        try {
+            if (headroot != null && Files.size(headroot) > 0l) {
+                List<String> lines = Files.readAllLines(headroot);
+                String line = lines.get(0);
+                String revLink = line.substring(line.indexOf(':')+1).trim();
+
+                Path revPath = headroot.getParent().resolve(revLink);
+                if(Files.isRegularFile(revPath) && Files.size(revPath) > 0l) {
+                    List<String> revlines = Files.readAllLines(revPath);
+                    String revline = revlines.get(0);
+                    if(revline.length()>=12){
+                        id = revline.substring(0, 12);
+                    } else {
+                        log("no content in " + revPath, Project.MSG_WARN);                        
```

Comment:
That log might be misleading. Including `revline` to the log adds precious context.
For instance :
```java
log("content of " + revPath + " could not be parsed as an ID : \"" + revline + "\"");
```

Reply:
The git-commit ID is longer than 12 characters (see git docs). so the "revline" is always longer, but has to be at least 12 characters.

Reply:
I don't see a benefit of another log line. the revline should always be a commit-id (again, see git docs) or something not git compliant.

Reply:
Agree with the "at least 12 chars". Actually I removed my comment about that before noticing your answer to it.
I dont suggest adding a new log, but adding context to the existing one.
The benefit is for the user who may have a corrupt (but non empty) "revPath" and would be puzzled by a "no content" log.